### PR TITLE
opt_frozen_dataclass: more improvements

### PR DESCRIPTION
**Pull request by matthiasdiener**
- raise an error when specifying `frozen`
- add tests
- add support for specifying `weakref_slot`

targets #273.

### CODE REVIEW

**Comment by matthiasdiener** [comment]
What do you think @inducer?

File: pytools/__init__.py [hunk 1]
```diff
@@ -3014,13 +3013,22 @@ def opt_frozen_dataclass(
...
+        raise ValueError("frozen must not be specified in opt_frozen_dataclass")
+
     def map_cls(cls: type[T]) -> type[T]:
         # This ensures that the resulting dataclass is hashable with and without
         # __debug__, unless the user overrides unsafe_hash or provides their own
         # __hash__ method.
+
+        # Make it possible to override 'frozen' in the class definition for testing.
+        # It would be nice to have something like https://discuss.python.org/t/allow-debug-to-be-set-at-runtime/64840
+        loc_frozen = __debug__ if "_frozen_override" not in cls.__dict__ else False
```
Comment (inducer) [inline pytools/__init__.py:3027]: Could you explain the use case?

Comment (inducer) [inline pytools/__init__.py:3027]: `frozen_override` could be in `kwargs` maybe?

File: pytools/test/test_dataclasses.py
Comment (inducer) [inline pytools/test/test_dataclasses.py:1]: I was thinking this needed tests. Thanks for adding them!

File: pytools/__init__.py [hunk 2]
```diff
@@ -2991,8 +2991,7 @@ def opt_frozen_dataclass(
             match_args: bool = True,
             kw_only: bool = False,
             slots: bool = False,
-            # Added in 3.11.
-            # weakref_slot: bool = False
+            **kwargs: Any,  # Extra, version dependent arguments (weakref_slot in 3.11)
```
Comment (inducer) [inline pytools/__init__.py:2994]: Is a `kwargs`-only `ParamSpec` a thing?

**Comment by inducer** [comment]
Thanks! I've rolled  most of this into #273, with attribution. We can tackle the override later on if it's still needed.

File: pytools/__init__.py (hunk 1 above)
Comment (matthiasdiener) [reply pytools/__init__.py:3027]: It's just meant for testing so that we can "fake" setting `__debug__=False`.

Comment (matthiasdiener) [reply pytools/__init__.py:3027]: Yeah, I think that's a good idea.

**Pull request closed without merging**